### ISSUE TITLE
Calling MKL function before MPI_Init causes code to hang with MKL_CBWR=COMPATIBLE

### ISSUE THREAD

**Issue by ericch1** [open-mpi/ompi#5070]
Hi,

I have reproduced the bug with OpenMPI 2.1.2 and 3.0.1 and Intel MKL 2018.0 Update 1 Product build 20171007.

On the same machine, it works perfectly with MPICH.
On another machine with MKL 2015, the bug is not present at all.

It looks like a specific problem between MKL 2018 and OpenMPI...

Here is how to reproduce:

```
export MKL_CBWR=COMPATIBLE
export MKL_VERBOSE=1
mpicxx -L${MKLROOT}/lib/intel64 -lmkl_intel_lp64 -lmkl_core -lmkl_intel_thread -lmkl_blacs_intelmpi_lp64 -liomp5 -ldl -lpthread -o foo foo.cc
./foo
```

When bugging, it hangs at MPI_Init like this:

```
Foo is called 
MKL_VERBOSE Intel(R) MKL 2018.0 Update 1 Product build 20171007 for Intel(R) 64 architecture Intel(R) Architecture processors, Lnx 3.20GHz lp64 intel_thread NMICDev:0
MKL_VERBOSE DGETRF(27,27,0x1f34280,27,0x7ffd36f28c20,27) 22.90ms CNR:COMPATIBLE Dyn:1 FastMM:1 TID:0  NThr:6 WDiv:HOST:+0.000
Calling MPI_Init:
```

Thanks!

Eric

The foo.cc code is:
```c++
#include <cstdio>
#include  <mkl.h>
#include <mpi.h>
int foo() {
   printf("Foo is called\n");
   const int lN = 27;
   double* lA = (double*)malloc(lN*lN*sizeof(double));
   for (int i = 0; i < lN*lN; ++i) {
     lA[i] = i;
   }
   int lPiv[lN];
   int lRes;
   dgetrf_(&lN, &lN, lA, &lN, lPiv, &lRes);
   return lRes;
}
class A
{
   public:
   A() { foo();}
};
static A sA; //A static declaration calls foo() !!

int main(int pArgc, char* pArgv[])
{
   printf("Calling MPI_Init:\n");
   MPI_Init(&pArgc, &pArgv);
   printf("MPI_Init done...\n");
   foo();
   MPI_Finalize();
   return 0;
}
```

**Comment by ericch1** [comment]
I submitted the issue to MKL forum too, since it is not clear for me if this is an OpenMPI or MKL related bug...

https://software.intel.com/node/776781

Eric

**Comment by AndreasKempfNEC** [comment]
Hi,

you are linking with `-lmkl_blacs_intelmpi_lp64`. As far as I understand, this means that the MKL makes use of Intel MPI. For other MPI implementations, you may have to use a different library as ABI incompatibilities can occur. Only implementations derived from MPICH might be [compatible](https://www.mpich.org/abi/) with each other. There is a description [here](https://software.intel.com/en-us/articles/using-intel-mkl-mpi-wrapper-with-the-intel-mkl-cluster-functions).

Andreas

**Comment by ericch1** [comment]
Hi Andreas,

thanks for your answer!  I never saw this before... After reading the doc, I was really hoping it will resolve my problem, so I gave it a try.
I did:
```
cd $MKLROOT/interfaces/mklmpi
make sointel64 interface=lp64 MPICC='/opt/openmpi-3.0.1/bin/mpicc' INSTALL_LIBNAME='libmkl_blacs_ompi3_lp64' INSTALL_DIR='/tmp' 
```
then I tried:
```
mpicxx -L/tmp -L${MKLROOT}/lib/intel64 -lmkl_intel_lp64 -lmkl_core -lmkl_gnu_thread -lmkl_blacs_ompi3_lp64 -liomp5 -ldl -lpthread  -o foo mkl_before_mpi_init.cc
```
And also with "-lmkl_intel_thread",
but they all failed the same way with this backtrace:
```
#0  0x00007ffff02c0e47 in sched_yield () from /lib64/libc.so.6
#1  0x00007ffff169ee74 in _INTERNAL_26_______src_z_Linux_util_cpp_d7ee2e5e::__kmp_atfork_prepare () at ../../src/z_Linux_util.cpp:1534
#2  0x00007ffff02a752d in fork () from /lib64/libc.so.6
#3  0x00007fffdebe2842 in rte_init.part () from /opt/openmpi-3.0.1/lib/openmpi/mca_ess_singleton.so
#4  0x00007fffeff4e8c6 in orte_init () from /opt/openmpi-3.0.1/lib/libopen-rte.so.40
#5  0x00007ffff0ece4ec in ompi_mpi_init () from /opt/openmpi-3.0.1/lib/libmpi.so.40
#6  0x00007ffff0ef6bdb in PMPI_Init () from /opt/openmpi-3.0.1/lib/libmpi.so.40
#7  0x0000000000400841 in main (pArgc=1, pArgv=0x7fffffffd5c8) at mkl_before_mpi_init.cc:26
```

As suggested by Intel guy, I upgraded to latest mkl release, but it changed nothing...

If you have other ideas, I will test them! :)

Thanks!

Eric

**Comment by ggouaillardet** [comment]
@ericch1 the hang occurs when your app, which is launched as a singleton, fork&exec orted, and the stack traces indicates it involves a hook in the intel runtime (aka `kmp`).

Can you please give the two workarounds a try ?

 - `mpirun -np 1 ./foo`
 - `OMPI_MCA_ess_singleton_isolated=true ./foo`

Unless your application will `MPI_Comm_spawn()`, you can use the `isolated` mode, and it will save you some time in `MPI_Init()` as well.

**Comment by ericch1** [comment]
Hi @ggouaillardet ,

both workaround works...
here is the output:
```
mpirun -n 1 ./foo 
Foo is called
MKL_VERBOSE Intel(R) MKL 2018.0 Update 2 Product build 20180127 for Intel(R) 64 architecture Intel(R) Architecture processors, Lnx 3.20GHz lp64 gnu_thread
MKL_VERBOSE DGETRF(27,27,0xba7280,27,0x7ffd7119f600,17) 2.38ms CNR:COMPATIBLE Dyn:1 FastMM:1 TID:0  NThr:1
Calling MPI_Init:
MPI_Init done...
Foo is called
MKL_VERBOSE DGETRF(27,27,0xcd39d0,27,0x7ffd7119f670,17) 43.04us CNR:COMPATIBLE Dyn:1 FastMM:1 TID:0  NThr:1
```
Knowing this do not occur with MKL2015, I am completely unable to say if this is a MKL bug or a bad MKL+OpenMPI mix or a bad use of either MKL or OpenMPI?

So, is there something to "fix" ?  Could this occur with another library?

Thanks,

Eric

**Comment by ggouaillardet** [comment]
At this stage, all I can say is that a `fork()` is involved, and if you run in __non__ `isolated` `singleton` mode, then Open MPI `fork()` under the hood.

I tried to check what `__kmp_atfork_prepare()` in published runtime sources, but the subroutines I found are empty.

can you try this program ?

```c
#include <cstdio>
#include  <mkl.h>
#include <stdlib.h>
int foo() {
   printf("Foo is called\n");
   const int lN = 27;
   double* lA = (double*)malloc(lN*lN*sizeof(double));
   for (int i = 0; i < lN*lN; ++i) {
     lA[i] = i;
   }
   int lPiv[lN];
   int lRes;
   dgetrf_(&lN, &lN, lA, &lN, lPiv, &lRes);
   return lRes;
}
class A
{
   public:
   A() { foo();}
};
static A sA; //A static declaration calls foo() !!

int main(int pArgc, char* pArgv[])
{
   printf("Calling system:\n");
   system("true");
   printf("system done...\n");
   foo();
   MPI_Finalize();
   return 0;
}
```

**Comment by ggouaillardet** [comment]
@rhc54 today, `ess/singleton` can be `isolated` or not.

That means we either never `fork()&exec()` `orted` (and hence `MPI_Comm_spawn()` and friend will fail when invoked), or we unconditionally `fork()&exec()` `orted` in `MPI_Init()` (and my gut feeling is we do not `MPI_Comm_spawn()` most of the time).

I am thinking a third scenario could be interesting. `fork()&exec()` `orted` when the first `MPI_Comm_spawn()` or friend is invoked.

Do you see anything that would make this approach difficult and/or virtually impossible ?

**Comment by rhc54** [comment]
Hmmm...there was a time when we actually did your third approach. IIRC, I hit an issue with it when we converted to PMIx - something to do with initialization, I think. You need the PMIx client library to initialize its internal storage without requiring connection to a server, and then re-initialize and connect to the server at some later point.

Should be doable, but may be a tad complex and requires some mods to PMIx itself. I have a vague feeling that I did something similar for tool connections (due to the race condition between the DVM's server starting up and the tool looking for it), so some groundwork may exist.

Bottom line: not impossible, but would probably take some effort and is not something you could backport.

Another option would be to do the interim storage at the OPAL level - i.e., call PMIx_Init, and if it fails (IIRC we return a unique error code indicating the server wasn't available), then use an OPAL-level hash table to store values. Try PMIx_Init each time someone calls opal_pmix.init.

Might be the easier approach, though it means some duplicate code in OPAL to handle the hash table. Would have to ensure that PMIx cleaned up completely each time PMIx_Init failed - we've never tested that.

**Comment by ericch1** [comment]
@ggouaillardet I compiled and launched your code (after commenting the MPI_Finalize call) and it works without any problem.

**Comment by ggouaillardet** [comment]
@ericch1 here is a simpler way to reproduce the issue, and it does not involve Open MPI whatsoever.

```c
#include <unistd.h>
#include <cstdio>
#include  <mkl.h>

int foo() {
   printf("Foo is called\n");
   const int lN = 27;
   double* lA = (double*)malloc(lN*lN*sizeof(double));
   for (int i = 0; i < lN*lN; ++i) {
     lA[i] = i;
   }
   int lPiv[lN];
   int lRes;
   dgetrf_(&lN, &lN, lA, &lN, lPiv, &lRes);
   return lRes;
}

int main(int pArgc, char* pArgv[])
{
   foo();
   printf("Forking:\n");
   pid_t pid = fork();
   if (0 == pid) {
       exit(0);
   }
   printf("Forked...\n");
   foo();
   return 0;
}
```

and then
```
$ g++ a.cpp -L${MKLROOT}/lib/intel64 -lmkl_intel_lp64 -lmkl_core -lmkl_gnu_thread -liomp5
$ MKL_CBWR=COMPATIBLE ./a.out 
```

Bottom line, `MKL` sets a hook before `fork()` that hangs.
A simple way to avoid this issue is to `vfork()` instead of `fork()`. These are not equivalent system calls, but as far as the `ess/singleton` component is concerned, `vfork()` is a fit imho.

Your initial test works for me with the following patch
```patch
diff --git a/orte/mca/ess/singleton/ess_singleton_module.c b/orte/mca/ess/singleton/ess_singleton_module.c
index 7729b9b..4b87d14 100644
--- a/orte/mca/ess/singleton/ess_singleton_module.c
+++ b/orte/mca/ess/singleton/ess_singleton_module.c
@@ -444,7 +444,7 @@ static int fork_hnp(void)
     opal_argv_append(&argc, &argv, "^s1,s2,cray,isolated");
 
     /* Fork off the child */
-    orte_process_info.hnp_pid = fork();
+    orte_process_info.hnp_pid = vfork();
     if(orte_process_info.hnp_pid < 0) {
         ORTE_ERROR_LOG(ORTE_ERR_SYS_LIMITS_CHILDREN);
         close(p[0]);
```

can you please give it a try and update the ticket you filled at Intel ?

**Comment by ericch1** [comment]
@ggouaillardet it is perfectly reproducing the bug! Thanks, I update the Intel ticket now.
Thanks a lot!

**Comment by ericch1** [comment]
On Intel forum, they proposed a workaround until this get fixed:
```
export KMP_INIT_AT_FORK=FALSE
```
and it works.
